### PR TITLE
papermc: 1.20.1.83 -> 1.20.2.223

### PR DESCRIPTION
**Pull request by Gerg-L**
## Description of changes
Bumped version and refactored to use `runCommandLocal` and `makeBinaryWrapper`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by eclairevoyant — COMMENTED** [review]
refactor and version bump should probably be separate commits. otherwise v nice :+1:

**Comment by Gerg-L** [comment]
> refactor and version bump should probably be separate commits

done

**Review by eclairevoyant — APPROVED** [review]

**Comment by wegank** [comment]
> refactored to use `runCommandLocal` and `makeBinaryWrapper`

The refactor doesn't make sense to me. Why did you think a package, not a trivial command, should be written with `runCommandLocal` instead?

**Comment by eclairevoyant** [comment]
is `iestall` + `makeBinaryWrapper` nontrivial?

**Comment by wegank** [comment]
I'd say look for `stdenvNoCC.mkDerivation`.

There are of course non-standard ways to make something buildable, for example, passing `postFetch` and `meta` into `fetchzip`. I just don't like to make exceptions for these patterns that are extremely rare in nixpkgs.

**Comment by Gerg-L** [comment]
`runCommandLocal` maps too `stdenvNoCC.mkDerivation` with a few flags so that's just un-abstracting a builder

But I get what you're saying I'll push changes tonight

File: pkgs/games/papermc/default.nix [hunk 1]
```diff
@@ -1,39 +1,52 @@
...
-  jar = let
-    mcVersion = lib.versions.pad 3 version;
-    buildNum = builtins.elemAt (lib.versions.splitVersion version) 3;
-  in fetchurl {
-    url = "https://papermc.io/api/v2/projects/paper/versions/${mcVersion}/builds/${buildNum}/downloads/paper-${mcVersion}-${buildNum}.jar";
-    sha256 = "sha256-HQpc3MOXa1wkXqgm9ciQj04FUIyuupnYiu+2RZ/sXE4=";
-  };
+  version = "1.20.2.234";
+in
+stdenvNoCC.mkDerivation {
```
Comment (wegank) [inline pkgs/games/papermc/default.nix:12]: A benefit of `mkDerivation` is that the following scheme ensures consistent overriding.

https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/stdenv.chapter.md#recursive-attributes-in-mkderivation-mkderivation-recursive-attributes

Basically, you need to put `pname` and `version` in `mkDerivation (finalAttrs: { })`, and use `finalAttrs.version` to refer to them in `src` or other places.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 256981` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>papermc</li>
  </ul>
</details>

**Pull request merged**